usecases/config/runtime/manager.go

```diff
@@ -0,0 +1,189 @@
+	// failures to load new configs.
+	currentConfig *T
+	currentHash   string
+	mu            sync.RWMutex // protects currentConfig
+
+	log             logrus.FieldLogger
+	lastLoadSuccess prometheus.Gauge
+	configHash      *prometheus.GaugeVec
+}
+
+func NewConfigManager[T any](
+	filepath string,
+	parser Parser[T],
+	interval time.Duration,
+	log logrus.FieldLogger,
+	r prometheus.Registerer,
+) (*ConfigManager[T], error) {
+	// catch empty filepath early
+	if len(strings.TrimSpace(filepath)) == 0 {
+		return nil, errors.New("filepath to load runtimeconfig is empty")
+	}
+
+	cm := &ConfigManager[T]{
+		path:     filepath,
+		parse:    parser,
+		interval: interval,
+		log:      log,
+		lastLoadSuccess: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_last_load_success",
+			Help: "Whether the last loading attempt of runtime config was success",
+		}),
+		configHash: promauto.With(r).NewGaugeVec(prometheus.GaugeOpts{
+			Name: "weaviate_runtime_config_hash",
+			Help: "Hash value of the currently active runtime configuration",
+		}, []string{"sha256"}), // sha256 is type of checksum and hard-coded for now
+	}
+
+	// try to load it once to fail early if configs are invalid
+	if err := cm.loadConfig(); err != nil {
+		return nil, err
+	}
+
+	return cm, nil
+}
+
+// Run is a blocking call that starts the configmanager actor. Consumer probably want to
+// call it in different groutine. It also respects the passed in `ctx`.
+// Meaning, cancelling the passed `ctx` stops the actor.
+func (cm *ConfigManager[T]) Run(ctx context.Context) error {
+	return cm.loop(ctx)
+}
+
+// Config returns the current valid config if available. Once the config manager
+// is started without any error, consumer should be able to get **valid** config
+// via this api.
+func (cm *ConfigManager[T]) Config() (*T, error) {
+	cm.mu.RLock()
+	defer cm.mu.RUnlock()
+
+	if cm.currentConfig == nil {
+		return nil, ErrEmptyConfig
+	}
+
+	return cm.currentConfig, nil
+}
+
+// loadConfig reads and unmarshal the config from the file location.
+func (cm *ConfigManager[T]) loadConfig() error {
+	f, err := os.Open(cm.path)
+	if err != nil {
+		cm.lastLoadSuccess.Set(0)
+		return errors.Join(ErrFailedToOpenConfig, err)
+	}
+	defer f.Close()
+
+	b, err := io.ReadAll(f)
+	if err != nil {
+		cm.lastLoadSuccess.Set(0)
+		return errors.Join(ErrFailedToReadConfig, err)
+	}
+
+	hash := fmt.Sprintf("%x", sha256.Sum256(b))
+	if hash == cm.currentHash {
+		cm.lastLoadSuccess.Set(1)
+		return nil // same file. no change
+	}
+
+	cfg, err := cm.parse(b)
+	if err != nil {
+		cm.lastLoadSuccess.Set(0)
+		return errors.Join(ErrFailedToParseConfig, err)
+	}
+
+	cm.updateConfig(cfg, hash)
+
+	return nil
+}
+
+func (cm *ConfigManager[T]) updateConfig(cfg *T, hash string) {
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
+	cm.currentConfig = cfg
+	cm.currentHash = hash
+
+	cm.lastLoadSuccess.Set(1)
+	cm.configHash.Reset()
+	cm.configHash.WithLabelValues(hash).Set(1)
```

Comment:
Could you extract metric setting out of the critical section as well? 

I'm nitpicking here a lot of as config managers tend to end up in a very hot places and it is better to have as little overhead as possible in these libraries.